jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java

```diff
@@ -340,44 +340,47 @@ public CaseDefinition getCase(String deploymentId, String caseDefinitionId) {
-        .map(n -> {
-            foundMilestones.add(n.getName());
-            return new CaseMilestoneInstanceImpl(String.valueOf(n.getId()), n.getName(), completedNodes.contains(n.getId()), n.getDataTimeStamp());        
-        })
-        .collect(toList());
-        
+
+        Collection<org.jbpm.services.api.model.NodeInstanceDesc> nodes = runtimeDataService.getValidNodeInstancesByCorrelationKeyNodeType(correlationKey,
+                                                                                                                                     Arrays.asList(ProcessInstance.STATE_ACTIVE, ProcessInstance.STATE_ABORTED),
+                                                                                                                                     Arrays.asList("MilestoneNode"),
+                                                                                                                                     queryContext);
+
+        // achieved milestones
+        Map<String, List<CaseMilestoneInstance>> milestonesGroup = nodes.stream()
+                                                                        .map(n -> (NodeInstanceDesc) n)
+                                                                        .filter(n -> n.getType() == 1)
+                                                                        .map(n -> new CaseMilestoneInstanceImpl(String.valueOf(n.getId()), n.getName(), true, n.getDataTimeStamp()))
+                                                                        .collect(Collectors.groupingBy(CaseMilestoneInstance::getName));
+
+        List<CaseMilestoneInstance> milestones = milestonesGroup.values().stream().map(e -> e.stream().min(this::compareMilestones).get()).collect(toList());
+        List<String> foundMilestones = milestones.stream().map(e -> e.getName()).collect(toList());
+
         if (!achievedOnly) {
             // add other milestones that are present in the definition
             CaseDefinition caseDef = getCase(pi.getDeploymentId(), pi.getProcessId());
             caseDef.getCaseMilestones().stream()
-            .filter(cm -> !foundMilestones.contains(cm.getName()))
-            .map(cm -> new CaseMilestoneInstanceImpl(cm.getId(), cm.getName(), false, null))
-            .forEach(cmi -> milestones.add(cmi));
+                   .filter(cm -> !foundMilestones.contains(cm.getName()))
+                   .map(cm -> new CaseMilestoneInstanceImpl(cm.getId(), cm.getName(), false, null))
+                   .forEach(cmi -> milestones.add(cmi));
         }
-        
-        return applyPagination(milestones, queryContext);
+
+        List<CaseMilestoneInstance> sortedMilestones = milestones.stream().sorted(this::compareMilestones).collect(toList());
+        return applyPagination(sortedMilestones, queryContext);
+    }
+
+    private int compareMilestones(CaseMilestoneInstance o1, CaseMilestoneInstance o2) {
+        if (o1.getAchievedAt() == null) {
+            return -1;
```

Comment:
if both are null, shouldn't it return 0?

Reply:
null <> null = true... null == null = false... depends on our view.